### PR TITLE
Allow actions in the new tab dropdown

### LINKED ISSUE #3759: Specify Panes in a Profile
# Description of the new feature/enhancement
The ability to save a tab with panes as a profile, or specify in a profile the panes to open when you open that profile.

I Have 6 projects I work on, on a daily basis at work, so I come in, open up ConEmu, and it opens a tab with 6 panes each set to a different git directory. Then I begin working.

So I want to make a `project` profile which opens up that setup. 

Currently being able to make panes is awesome, but I don't want to set it up every time I accidentally close the terminal or I reboot my computer. 

(I hope this matches the requirement of the template)

image

image

image

### LINKED ISSUE #9362: Make the Settings drop down able to switch back to the json
# Description of the new feature/enhancement
It appears the settings.json has become unbound and using the previous keybinding for it still opens the Settings UI. I can add a custom keybinding for the settings.json but I usually am lazy and simply click on the Settings drop down button as it's quicker and easier for me.  So I'd like the ability to switch out the default experience for the Settings drop down. 

# Proposed technical implementation details (optional)
Add a global setting that allows the user to switch out the default experience. This should be able to be set from both the Settings UI and the settings.json.

### PR DESCRIPTION
**Pull request by PankajBhojwani**
Allows the user to define entries in the new tab menu that execute actions, based on their action Id

Closes #3759
Closes #9362

### CODE REVIEW

**Comment by DHowett** [comment]
YAAAAAAAAAAAAAAASS

**Comment by zadjii-msft** [comment]
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 1]
```diff
@@ -1094,6 +1107,42 @@ namespace winrt::TerminalApp::implementation
         return profileMenuItem;
     }
 
+    // Method Description:
+    // - This method creates a flyout menu item for a given action
+    //   It makes sure to set the correct icon, keybinding, and click-action.
+    WUX::Controls::MenuFlyoutItem TerminalPage::_CreateNewTabFlyoutAction(const winrt::hstring& actionId)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1113]: ~nit: we could probably just pass the action in directly here~

eh, we need to look up the keybinding here too. I take it back, this is fine.

**Review by zadjii-msft — APPROVED** [review]
this was so easy and nice to review

**Comment by zadjii-msft** [comment]
may want to add a "closes #3759"

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 2]
```diff
@@ -1004,6 +1004,19 @@ namespace winrt::TerminalApp::implementation
...
+            {
+                const auto actionEntry = entry.as<ActionEntry>();
+                const auto actionId = actionEntry.ActionId();
+                if (_settings.ActionMap().GetActionById(actionId) == nullptr)
+                {
+                    break;
+                }
+
+                auto actionItem = _CreateNewTabFlyoutAction(actionId);
+                items.push_back(actionItem);
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1017]:
```suggestion
                if (_settings.ActionMap().GetActionById(actionId) )
                {
                    break;
                    auto actionItem = _CreateNewTabFlyoutAction(actionId);
                    items.push_back(actionItem);
                }
```
nit: you can consolidate this a bit

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 3]
```diff
@@ -1094,6 +1107,42 @@ namespace winrt::TerminalApp::implementation
...
+        auto actionMenuItem = WUX::Controls::MenuFlyoutItem{};
+        const auto action{ _settings.ActionMap().GetActionById(actionId) };
+        const auto actionKeyChord{ _settings.ActionMap().GetKeyBindingForAction(actionId) };
+
+        if (actionKeyChord)
+        {
+            _SetAcceleratorForMenuItem(actionMenuItem, actionKeyChord);
+        }
+
+        auto actionName = action.Name();
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1124]:
```suggestion
        const auto actionName = action.Name();
```

**Review by carlos-zamora — APPROVED** [review]
+1 to https://github.com/microsoft/terminal/pull/17281#issuecomment-2137185547 before you merge

Great work!

**Comment by DHowett** [comment]
> may want to add a
> 
> closes #3759 
> closes #9362

Very explicitly, you need to put them on separate lines both of which start with "Closes".

File: src/cascadia/TerminalSettingsModel/ActionMap.cpp [hunk 4]
```diff
@@ -599,6 +599,11 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         return _GetActionByKeyChordInternal(keys).value_or(nullptr);
     }
 
+    Model::Command ActionMap::GetActionById(const winrt::hstring& cmdID) const
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionMap.cpp:602]: everywhere else we call this `ID` - why is it `Id` here?

**Comment by PankajBhojwani** [comment]
## Validation Steps Performed
newTabMenuAction

File: src/cascadia/TerminalSettingsModel/ActionEntry.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#include "pch.h"
+#include "ActionEntry.h"
+#include "JsonUtils.h"
+
+#include "ActionEntry.g.cpp"
+
+using namespace Microsoft::Terminal::Settings::Model;
+using namespace winrt::Microsoft::Terminal::Settings::Model::implementation;
+
+static constexpr std::string_view ActionIdKey{ "actionId" };
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionEntry.cpp:13]: qq: why should we stutter?

```
{ "type": "action", "actionId": "Terminal.Quit" }
           ^^^^^^    ^^^^^^
                \---/
                  `they are the same word
```

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/ActionEntry.cpp:13]: hearing arguments on whether it should/could just be "id"

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/ActionEntry.cpp:13]: FWIW only the json field need change if we decide to call it "id". We don't need to rename the member or anything.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 6]
```diff
@@ -1094,6 +1106,42 @@ namespace winrt::TerminalApp::implementation
...
+        const auto action{ _settings.ActionMap().GetActionById(actionId) };
+        const auto actionKeyChord{ _settings.ActionMap().GetKeyBindingForAction(actionId) };
+
+        if (actionKeyChord)
+        {
+            _SetAcceleratorForMenuItem(actionMenuItem, actionKeyChord);
+        }
+
+        const auto actionName = action.Name();
+        actionMenuItem.Text(actionName);
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1124]: any reason we need to save this as a local? we might save one string ref count if we just do... `Text(action.Name())`

**Review by DHowett — CHANGES_REQUESTED** [review]
open discussion on `id` vs `actionId`!

File: src/cascadia/TerminalSettingsModel/ActionEntry.cpp (hunk 5 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/ActionEntry.cpp:13]: aw thanks

**Review by zadjii-msft — APPROVED** [review]

**Review by DHowett — DISMISSED** [review]

**Review by DHowett — APPROVED** [review]

**Review by zadjii-msft — APPROVED** [review]

**Comment by erikzhang** [comment]
Can I assign an icon to a command?

**Comment by PankajBhojwani** [comment]
> Can I assign an icon to a command?

Yep!

**Pull request merged**